pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,178 @@
+	return &FileWatcherOptions{
+		getProcessPIDByNameFn: GetProcessPIDByName,
+		signalFn: func(pid int) error {
+			return syscall.Kill(pid, syscall.SIGTERM)
+		},
+	}
+}
+
+func NewFileWatcherWatchdog() *cobra.Command {
+	o := NewFileWatcherOptions()
+
+	cmd := &cobra.Command{
+		Use:   "file-watcher-watchdog",
+		Short: "Watch files on the disk and kill the specified process on change",
+		Run: func(cmd *cobra.Command, args []string) {
+			klog.V(1).Info(cmd.Flags())
+			klog.V(1).Info(spew.Sdump(o))
+
+			ctx := context.TODO()
+
+			if err := o.Complete(); err != nil {
+				klog.Fatal(err)
+			}
+			if err := o.Validate(); err != nil {
+				klog.Fatal(err)
+			}
+
+			if err := o.Run(ctx); err != nil {
+				klog.Fatal(err)
+			}
+		},
+	}
+
+	o.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (o *FileWatcherOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ProcessName, "process-name", "", "name of the process to send signal to on file change.")
+	fs.StringSliceVar(&o.Files, "files", o.Files, "list of file names to react on")
+}
+
+func (o *FileWatcherOptions) Complete() error {
+	o.initialContent = map[string][]byte{}
+	for _, name := range o.Files {
+		content, err := ioutil.ReadFile(name)
+		if err != nil {
+			return err
+		}
+		o.initialContent[name] = content
+	}
+	return nil
+}
+
+func (o *FileWatcherOptions) Validate() error {
+	if len(o.ProcessName) == 0 {
+		return fmt.Errorf("process name must be specified")
+	}
+	if len(o.Files) == 0 {
+		return fmt.Errorf("at least one file to observe must be specified")
+	}
+	return nil
+}
+
+// monitorProcessPID will get the PID of the process we monitor.
+// In only gets the PID once. When the PID change for the process, it shutdown the watchdog
+// in reaction to reset the file observers.
+func (o *FileWatcherOptions) monitorProcessPID(shutdown context.CancelFunc, pidObservedCh chan struct{}) {
+	defer shutdown()
+	for {
+		// get the monitored process PID
+		pid, found, err := o.getProcessPIDByNameFn(o.ProcessName)
+		if !found {
+			klog.Warningf("Process %q does not have PID", o.ProcessName)
+			return
+		}
+		if err != nil {
+			klog.Warning(err)
+			return
+		}
+
+		if o.getCurrentPID() == 0 {
+			o.setCurrentPID(pid)
+			close(pidObservedCh)
+		}
+
+		// when the PID change, signal shutdown
+		if o.getCurrentPID() != pid {
+			return
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (o *FileWatcherOptions) setCurrentPID(pid int) {
+	o.Lock()
+	defer o.Unlock()
+	o.currentProcessPID = pid
+}
+
+func (o *FileWatcherOptions) getCurrentPID() int {
+	o.Lock()
+	defer o.Unlock()
+	return o.currentProcessPID
+}
+
+func (o *FileWatcherOptions) Run(ctx context.Context) error {
+	runCtx, shutdown := context.WithCancel(ctx)
+
+	pidObservedCh := make(chan struct{})
+	go o.monitorProcessPID(shutdown, pidObservedCh)
+
+	// Wait while we get the initial PID for the process
+	klog.Infof("Waiting for process %q ...", o.ProcessName)
+	<-pidObservedCh
```

Comment:
this is the sweet spot... we can prefix all Files with `/proc/getCurrentPID()/root/` :-)

Reply:
when the PID change, the watchdog will shutdown (and container will restart)... it will get the new PID and prefix the files again.